### PR TITLE
Add -iree-vulkan-keep-shader-modules flag.

### PR DESCRIPTION
**Pull request by ScottTodd**
These can be disassembled with `spirv-dis` (see https://github.com/KhronosGroup/SPIRV-Tools) or used by other tools/programs.

### CODE REVIEW

File: iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp [hunk 1]
```diff
@@ -343,6 +350,32 @@ class VulkanSPIRVTargetBackend : public TargetBackend {
         configAttr);
   }
 
+  void saveSpirvBinary(IREE::HAL::ExecutableVariantOp variantOp,
+                       ArrayRef<uint32_t> binary) {
+    llvm::SmallString<32> filePath;
+    if (std::error_code error = llvm::sys::fs::createTemporaryFile(
+            variantOp.getName(), "spvasm", filePath)) {
+      llvm::errs() << "failed to generate file for spirv binary: "
```
Comment (antiagainst) [inline iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp:358]: Nit: s/spirv/SPIR-V/

Same for the rest. :)

File: iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp [hunk 2]
```diff
@@ -343,6 +350,32 @@ class VulkanSPIRVTargetBackend : public TargetBackend {
...
+    auto file = std::make_unique<llvm::ToolOutputFile>(filePath, error,
+                                                       llvm::sys::fs::OF_None);
+    if (error) {
+      llvm::errs() << "failed to open file for spirv binary '" << filePath
+                   << "': " << error.message();
+      return;
+    }
+
+    mlir::emitRemark(variantOp.getLoc())
+        << "spirv binary for " << variantOp.getName() << " preserved:\n"
```
Comment (antiagainst) [inline iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp:372]: Nit: quote the name with `''`? (I recall we have wired names that itself contains double quotes and other special chars.. Which is not very friendly to OS in general. A separate issue though.)

**Review by antiagainst — APPROVED** [review]
Nice, thanks!

File: iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp (hunk 2 above)
Comment (ScottTodd) [reply iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp]: This matches the other code here: https://github.com/google/iree/blob/c65709ec69b6b70865b626d43dcfa785b1c596bb/iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp#L492-L496

It's using a temp path and starts on a newline, so it's pretty readable without quotes:
```mlir
D:\dev\projects\iree/iree/test/e2e/models/unidirectional_lstm.mlir:149:8: remark: spirv binary for vulkan_spirv_fb preserved:
    C:\Users\Scott\AppData\Local\Temp\vulkan_spirv_fb-6b33a8.spv
  %3 = "mhlo.pad"(%2, %cst) {edge_padding_high = dense<[0, 0, 60]> : tensor<3xi64>, edge_padding_low = dense<0> : tensor<3xi64>, interior_padding = dense<0> : tensor<3xi64>} : (tensor<1x5x4xf32>, tensor<f32>) -> tensor<1x5x64xf32>
       ^
D:\dev\projects\iree/iree/test/e2e/models/unidirectional_lstm.mlir:37:8: remark: spirv binary for vulkan_spirv_fb preserved:
    C:\Users\Scott\AppData\Local\Temp\vulkan_spirv_fb-a6ad47.spv
  %9 = "mhlo.reduce"(%arg3, %cst) ( {
       ^
```

Comment (antiagainst) [reply iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp]: Oh, I thought it's the dispatch name.. This makes sense then. Thanks for the clarification!

**Pull request merged**